### ISSUE TITLE
Allow non-ASCII characters in the status line

### ISSUE THREAD

**Issue by v-python** [bottlepy/bottle#1187]
This is one message I often see, from running a bottle server. I'm not an expert in either bottle or wsgi, but migrated from using the default http.server to bottle to get some extra capability, and generally it works well, while developing some web applications. However, a user recently sent the following traceback from the bottle console, and it ends with the subject message.

I note that the first thing that went wrong was the user hit a bug in my server application, which raised the "Missing data" message.  Note that the data in the "Missing data" message includes a Unicode subscript 4 codepoint, but it was displayed fine on the bottle console. (Python 3.something learned how to write Unicode on the console, 3.0 didn't know how, but its been a few years now.)

All the stuff after the "Missing data" message is apparently generated by bottle, while handling the raised error. It appears, from a novice's point of view, that there are a number of issues in the bottle and/or wsgi error handling process.

1. The UnicodeEncodeError: I don't know what is trying to be encoded, but it would appear that it includes the Unicode subscript 4 (U+2084), which indicates that it is trying to include user data (my application data) in whatever is trying to be encoded. I think it is rather incorrect to assume that user data can be encoded as latin-1. The attempt to do so obviously fails!

2. The next UnicodeEncodeError seems to have the same traceback except that it adds a line "Exception happened during processing of a request...". So I'm not sure of the benefit of reporting it twice, but I suppose that is safer than not reporting it at all.

3. Then we get a stack trace culminating in "TypeError: 'NoneType' object is not subscriptable. I don't think this is a direct result of a UnicodeEncodeError, but perhaps because of that error occurring where it did (send_headers, send_preamble), the code that calls send_preamble doesn't properly handle the possibility of getting an exception there?  I see that message after other errors also, like "ConnectionResetError", which also happens with a stack trace includes send_headers.

4. Finally, we get a stack trace culminating in "AttributeError: 'NoneType' object has no attribute 'split'". Perhaps it is another consequence of the earlier error(s), but again, if errors can happen, the cleanup code shouldn't itself produce more errors.

My question is whether these are bugs in bottlepy, misusing wsgi in some way, or in wsgi. I'm happy to attempt to gather more information if I can, and am told how. I'm happy to open a ticket against the python stdlib if the problem is in wsgiref. But I don't understand it enough at the moment to open a sensible ticket there, since I'm not sure of the protocol and responsibilities between bottle and wsgi.

The rest of this is the data from the bottle console:

"Missing data for chord edit op 0: {'C': 11, 'R': 'z3-', 'J': {'T': 'n', 'E': [{'S': 'c', 'O': 2, 'x': 17.2, 'l': 87, 'd': 'up', 'f': 'None', 'N': [{'O': 0, 'v': 'S', 'n': 'E₄', 'd': 32, 's': 'left', 'u': 0, 'y': 0, 'ax': 0, 'dx': 0, 'dy': 0, 'r': 0, 'lsd': 'up', 'lsx': 0, 'lsy': 2, 'rsd': 'None', 'rsx': 0, 'rsy': 0}]}], 'a': 0}}"
127.0.0.1 - - [17/Dec/2019 12:51:54] "PATCH /Chinese-proof/Chinese-20191216-engraver/patch/c~64~ HTTP/1.1" 400 412
Traceback (most recent call last):
  File "C:\Users\richa\Dropbox\0-gp\py32\python-3.7.2\lib\wsgiref\handlers.py", line 138, in run
    self.finish_response()
  File "C:\Users\richa\Dropbox\0-gp\py32\python-3.7.2\lib\wsgiref\handlers.py", line 180, in finish_response
    self.write(data)
  File "C:\Users\richa\Dropbox\0-gp\py32\python-3.7.2\lib\wsgiref\handlers.py", line 274, in write
    self.send_headers()
  File "C:\Users\richa\Dropbox\0-gp\py32\python-3.7.2\lib\wsgiref\handlers.py", line 332, in send_headers
    self.send_preamble()
  File "C:\Users\richa\Dropbox\0-gp\py32\python-3.7.2\lib\wsgiref\handlers.py", line 252, in send_preamble
    self._write(('HTTP/%s %s\r\n' % (self.http_version,self.status)).encode('iso-8859-1'))
UnicodeEncodeError: 'latin-1' codec can't encode character '\u2084' in position 187: ordinal not in range(256)
127.0.0.1 - - [17/Dec/2019 12:51:54] "PATCH /Chinese-proof/Chinese-20191216-engraver/patch/c~64~ HTTP/1.1" 500 59

Exception happened during processing of request from ('127.0.0.1', 51608)
Traceback (most recent call last):
  File "C:\Users\richa\Dropbox\0-gp\py32\python-3.7.2\lib\wsgiref\handlers.py", line 138, in run
    self.finish_response()
  File "C:\Users\richa\Dropbox\0-gp\py32\python-3.7.2\lib\wsgiref\handlers.py", line 180, in finish_response
    self.write(data)
  File "C:\Users\richa\Dropbox\0-gp\py32\python-3.7.2\lib\wsgiref\handlers.py", line 274, in write
    self.send_headers()
  File "C:\Users\richa\Dropbox\0-gp\py32\python-3.7.2\lib\wsgiref\handlers.py", line 332, in send_headers
    self.send_preamble()
  File "C:\Users\richa\Dropbox\0-gp\py32\python-3.7.2\lib\wsgiref\handlers.py", line 252, in send_preamble
    self._write(('HTTP/%s %s\r\n' % (self.http_version,self.status)).encode('iso-8859-1'))
UnicodeEncodeError: 'latin-1' codec can't encode character '\u2084' in position 187: ordinal not in range(256)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\richa\Dropbox\0-gp\py32\python-3.7.2\lib\wsgiref\handlers.py", line 141, in run
    self.handle_error()
  File "C:\Users\richa\Dropbox\0-gp\py32\python-3.7.2\lib\wsgiref\handlers.py", line 368, in handle_error
    self.finish_response()
  File "C:\Users\richa\Dropbox\0-gp\py32\python-3.7.2\lib\wsgiref\handlers.py", line 180, in finish_response
    self.write(data)
  File "C:\Users\richa\Dropbox\0-gp\py32\python-3.7.2\lib\wsgiref\handlers.py", line 274, in write
    self.send_headers()
  File "C:\Users\richa\Dropbox\0-gp\py32\python-3.7.2\lib\wsgiref\handlers.py", line 331, in send_headers
    if not self.origin_server or self.client_is_modern():
  File "C:\Users\richa\Dropbox\0-gp\py32\python-3.7.2\lib\wsgiref\handlers.py", line 344, in client_is_modern
    return self.environ['SERVER_PROTOCOL'].upper() != 'HTTP/0.9'
TypeError: 'NoneType' object is not subscriptable

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\richa\Dropbox\0-gp\py32\python-3.7.2\lib\socketserver.py", line 650, in process_request_thread
    self.finish_request(request, client_address)
  File "C:\Users\richa\Dropbox\0-gp\py32\python-3.7.2\lib\socketserver.py", line 360, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "C:\Users\richa\Dropbox\0-gp\py32\python-3.7.2\lib\socketserver.py", line 720, in __init__
    self.handle()
  File "C:\Users\richa\Dropbox\0-gp\py32\python-3.7.2\lib\wsgiref\simple_server.py", line 133, in handle
    handler.run(self.server.get_app())
  File "C:\Users\richa\Dropbox\0-gp\py32\python-3.7.2\lib\wsgiref\handlers.py", line 144, in run
    self.close()
  File "C:\Users\richa\Dropbox\0-gp\py32\python-3.7.2\lib\wsgiref\simple_server.py", line 35, in close
    self.status.split(' ',1)[0], self.bytes_sent
AttributeError: 'NoneType' object has no attribute 'split'
----------------------------------------

**Comment by Marrin** [comment]
> I think it is rather incorrect to assume that user data can be encoded as latin-1. The attempt to do so obviously fails!

It happens in generating the HTTP response line and that part of the response is by definition/specification encoded in Latin-1.  The actual value that has the problem, the path portion of the URL is even restricted to ASCII with stuff outside of ASCII encoded in some encoding and then URL encoded/percent encoded. Browsers usually encode as UTF-8 before percent encoding.

Could you write and show a simple bottle app, describe how to start it, and then how to trigger this error with `HTTPie`, `wget`, or `curl`?

**Comment by v-python** [comment]
I'll give it a shot.

From what you say, it sounds like if user data is included in an area that is restricted to Latin-1, then it should be properly encoded into Latin-1 with escapes, rather than being assumed to be Latin-1.

I know about the URL restrictions; I'm less clear on what parts of the response might have other restrictions. Without the error condition, the request is properly handled... so it seems that the error is in the error path, not the success path.

**Comment by v-python** [comment]
curl -i -X GET -H "Content-Type:application/json" -d "{\"a\":\"b\"}" http://localhost:8080/raise

to this bottle server... Produces the UnicodeEncodeError, but not the rest. But this is a start.

`__version__ = '19.12.17'

import json
from bottle import Bottle, request, response, HTTPResponse, \
    static_file, redirect, abort, debug

app = Bottle()

#####################################
def processpatch( dir, fn, payload ):  # version 19.9.1
    global lastspbdir, lastspb
    if dir == lastspbdir:
        spb = lastspb
    else:
        spb = SPBInterface( dir )
        lastspb = spb
        lastspbdir = dir

    logpath, logfn = posixpath.split( fn )
    logfn = posixpath.join( logpath, 'log', logfn + '.log')
    with open( logfn, 'a', encoding='UTF-8') as logfh:
        logfh.write( f'''¶ {request.environ.get('SERVER_NAME')} @ {email.utils.formatdate(time.time(), usegmt=True)} payload: {payload}\n''')
    return spb.dopatchop( fn, payload )

@app.patch('/<host>/<path:path>/patch/<fil:path>')
def patchcmd( host, path, fil ):  # version 19.4.12
    fn = posixpath.join( path, fil )
    payload = request.json
    respl = processpatch( path, fn, payload )
    response.headers['Content-Type'] = 'application/json'
    response.status = respl.status
    return json.dumps( respl )

@app.route('/raise')
def files():
    payload = request.json
    respl = {}
    print( payload )
    try:
        raise ValueError( f'raised ₄')
    except Exception as exc:
        msg = str( exc )
        print( msg )
        response.status = f'400 {msg}'
    return json.dumps( msg )

@app.get('<path>')
def host( path ):
    return 'Hello, World'

app.run()
`

**Comment by v-python** [comment]
I told it insert code... and pasted between the inserted quotes.  What does it need to preserve formatting?  I'm not github expert either.

**Comment by Marrin** [comment]
@v-python Looks like an error on your side as you set `response.status_line` to something that is not encodable as latin-1.  Personally I've never changed the status line, just the status code.  For additional information I use the response body.  And there's more than the encoding to look out for — also newline and carriage return characters should not be in the status line.  I guess there's a length limit too.

A simple fix would be to force the formatted message to ASCII only:

```python
In [8]: message                                                                 
Out[8]: 'raised ₄'

In [9]: f"400 {message!a}"                                                      
Out[9]: "400 'raised \\u2084'"
```

I would be careful with just converting _any_ exception to a string and put that into a response as this might leak information that could be used to hack into the web application.  Unexpected exceptions should usually be kept internally as _500 Internal Server Error_ with the details logged to the server logs.

The _insert code_ button is for inline code like `response.status_line` above, not for a code block.  Code blocks are either indented, or ”fenced” by three backticks before and after the code without extra indention. See [Github Flavored Markdown](https://guides.github.com/features/mastering-markdown/#GitHub-flavored-markdown).

**Comment by defnull** [comment]
RFC7230 allows a lot more in the reason-phrase part of the status-line than you'd expect, but HTTP is still a binary protocol and unicode must be encoded to bytes at some point. WSGI (PEP-3333) dodges the question a bit by requiring all unicode strings passed to `start_response` to be `latin1` encode-able. To support unicode in headers (e.g. cookies), bottle already re-encodes header values (`value.encode('utf8').decode('latin1')`) before they are passed to `start_response()`. When encoded with `latin1` again, this results in valid `utf8` bytes on the wire. Bottle does not do that (yet) for the status line. Probably because it is rather uncommon to actually set the whole line instead of just the code.

While non-ASCII characters in a status line are uncommon and not really recommended, it would be valid, according to RFC7230. So, yes, this is a bug in bottle (but not a very urgent one). I'll fix that in 0.13.

It is still a bad idea to do that in the first place. Perhaps just do not do that?

**Comment by v-python** [comment]
@Marrin thanks for the helpful comments. Presently the application is only usable via localhost, so exposure of data to the single computer/user using it is not problematical.

@defnull Thanks for the future-fix.  If I figure out a reproducible case for the other errors, I'll open another ticket.
